### ISSUE TITLE
`SmagorinskyLilly` and `AnisotropicMinimumDissipation` unrealistic in 2D

### ISSUE THREAD

**Issue by glwagner** [iuryt/LockExchange.jl#5]
I noticed the experiment is 2D:

https://github.com/iuryt/LockExchange.jl/blob/8e9f73a28881803ca6e43ab2d515b8ca351964e3/code/lockexchange.jl#L15

but uses `SmagorinskyLilly`:

https://github.com/iuryt/LockExchange.jl/blob/8e9f73a28881803ca6e43ab2d515b8ca351964e3/code/lockexchange.jl#L21

careful about this, because LES closures are really only going to be valid in 3D! That said, it would be interesting to _see_ the differences --- maybe you can illustrate those here!

**Comment by iuryt** [comment]
@AgataPB  can you run the same experiment with `AnisotropicMinimumDissipation()` instead of `SmagorinskyLilly()` and post both GIFs here? 

John Taylor also mentioned about using a different advection scheme. Maybe `WENO()`?

**Comment by iuryt** [comment]
@glwagner what do you suggest? Adding some points in `X`?

**Comment by glwagner** [comment]
> Maybe WENO()?

`advection = WENO5()` is the best we have. You can also try `UpwindBiasedFifthOrder()`, which is a bit faster, especially on the CPU.

We need to change the default to `UpwindBiasedFifthOrder()`... more or less, I wouldn't ever use anything other than those two.

> Adding some points in X?

Either make it 3D (which will be expensive) or you could also just use `ScalarBiharmonicDiffusivity` instead. Unfortunately you then need to choose viscosity and diffusivity. You can estimate an appropriate value based on the grid spacing to the fourth power (the length scale) and some characteristic dissipation time-scale (eg some multiple of the time-step).

With `WENO5()` you _may_ even get away with `closure=nothing`, if your time steps are small enough.

It'd be interesting to see what `AnisotropicMinimumDissipation()` does anyways. Very likely `SmagorinskyLilly()` is too diffusive (this would be true even in 3D, but maybe exacerbated further in 2D --- not sure).

**Comment by AgataPB** [comment]
Hi! 
I have changed for the advection = WENO5() and it has improved the results in terms of the lower than initial conditions temperatures on the denser layer that we are getting. The UpwindBiasedFifthOrder() improved this also but the WENO5 was better.   
Now, when I changed the closure it did not change the results much. The results are down below:

Default advection scheme and SmagorinskyLilly closure:
![animation](https://user-images.githubusercontent.com/37347804/158224753-c018d913-3b29-4a99-b247-e6693384ff35.gif)


First example: WENO5 and No closure
![WENO5_NoClosure](https://user-images.githubusercontent.com/37347804/158223357-8f5373cd-d19b-410e-845e-fc28ced76820.gif)

Second: WENO5 and SmagorinskyLilly closure
![WENO5_SmagoClosure](https://user-images.githubusercontent.com/37347804/158223260-8f13103b-3343-4d2e-a432-6e3ac610bd15.gif)

Third: WENO5 and AnisotropicMinimumDissipation closure
![WENO5_AnisoClosure](https://user-images.githubusercontent.com/37347804/158223522-a9ea8adb-8c1e-48fa-a165-85f32169805e.gif)



Making it 3D (Nx = 2, total 1m width, is it enough? ) I got:

First example: 3D WENO5 and No closure
![3D_WENO5_NoClosure](https://user-images.githubusercontent.com/37347804/158249721-aa09fc3d-de81-408c-bed1-961667f5dfb5.gif)


Second: 3D WENO5 and SmagorinskyLilly closure
![3DWENO5_SmagoClosure](https://user-images.githubusercontent.com/37347804/158242397-7a3699b3-a7d2-4e99-a95a-30e682694472.gif)


Third: 3D WENO5 and AnisotropicMinimumDissipation closure
![3D_WENO5_AnisoClosure](https://user-images.githubusercontent.com/37347804/158246632-ba1cc73b-c56d-406f-9158-c1ac3c03c534.gif)




For the ScalarBiharmonicDiffusivity, I got NANs, both with 2D and 3D. I believe I am estimating the viscosity and diffusivity wrong. What I did was:
ν = 0.1^4/ 0.025seconds  # [m⁴ s⁻¹]
closure = ScalarBiharmonicDiffusivity(ν=ν, κ=ν) 
I also decreased the time step but it did not work --`.

**Comment by iuryt** [comment]
@AgataPB what is the topology of the x-axis on the 3D simulations? Periodic?
Are you showing the first slice or the x-averaged field?

I would say we need enough points in x to have at least two stencil length for the advection scheme. I am not familiar with WENO methods, but I believe WENO5 stands for a fifth-order WENO scheme, so then we would need at least 10 points?

It took me a while to stop looking those mesmerizing animations! haha

**Comment by tomchor** [comment]
Sorry to drop in the discussion. First of all, nice simulation and interesting discussion!

> Making it 3D (Nx = 2, total 1m width, is it enough? ) I got:

I think the reason why LES closures aren't recommended for 2D simulations is not due to numerics, but physics. Basically subgrid-scale closures for LES make assumptions that are only valid in 3D turbulence (chief of which is the forward energy cascade). So basically you'd need `Lx` to be large enough that the turbulence can act "isotropically". You can run your 2D simulation with Smag or AMD and it might give you a reasonable result, but I personally wouldn't take that simulation too seriously.

PS: It's super interesting to see the effect of the default advection scheme (which I think is 3rd order upwind?) compared to WENO5. This is a especially tricky set-up because of the sharp initial gradient, but still kinda surprising to see how much better WENO5 is. That anomaly at the bottom with the default scheme is crazy...

**Comment by iuryt** [comment]
I think that to keep it simple we can run 2D with `WENO5` and no closure. We might need to use a `tanh` instead of a step function and tweak the gradient to avoid having a too-small `dt` and making this too expensive. I am still intrigued on how it looks like with `UpwindBiasedFifthOrder`.

**Comment by glwagner** [comment]
Keeping 2D is a good idea! As @tomchor implied you need more than 2 points to resolve 3D dynamics. Probably no fewer than 32 for LES, and more would be better (depending on the problem). It’s still sweet in 2D, even if idealized!

**Comment by johnryantaylor** [comment]
I agree with @tomchor.  Using an explicit LES subgrid-scale model (SmagLilly, AMD, etc.) in a 2D simulation isn't particularly meaningful since these models are formulated based on assumptions about 3D turbulence.  Hyper-viscosity/diffusivity might be a good choice for 2D in this case.  That should give you a bit of extra dissipation around the nose of the gravity current to prevent numerical issues without damping the rest of the flow too much.

**Comment by johnryantaylor** [comment]
Also, it looks like some of the oscillations in density are generated right at the start of the simulation where the density looks like a step function.  Starting with a smoother function (e.g. erf or tanh) might help.

**Comment by AgataPB** [comment]
> @AgataPB what is the topology of the x-axis on the 3D simulations? Periodic? Are you showing the first slice or the x-averaged field?
> 
> I would say we need enough points in x to have at least two stencil length for the advection scheme. I am not familiar with WENO methods, but I believe WENO5 stands for a fifth-order WENO scheme, so then we would need at least 10 points?
> 
> It took me a while to stop looking those mesmerizing animations! haha


I used Bounded topology for x too. And it was only the first slice (but second is almost the same as the first).

I will try to make a more robust (more points) 3D next week! I will run in the MGHPCC. I was shocked how well these experiments run in my 2015 laptop, ha! Super cool o-o
